### ISSUE TITLE
Non-class based SC engines

### ISSUE THREAD

**Issue by antonhornquist** [monome/norns#863]
Let's have an issue to discuss engines based on interpreted SC code.

The problem:
If `.sc` files in the SuperCollider Extensions folder contains duplicate classes SCLang will not boot and no engines can be used.

Solution:
Change (*breaking change*) so SC engines are written in interpreted SClang which better supports coexisting engines.

Other things.
- Supporting ugens & classes
Do we need to cater for custom ugens or SC classes to be used? If so, they are to some extens subject to name clashes. Ie. all classes cannot be included by default in the Extensions folder or we will still have the SC duplicate class problem. @catfact has discussed introducing an `install` step for this.

**Comment by antonhornquist** [comment]
The engine architecture, on a high level, is simple:
Each engine needs a way to get (1) allocated and (2) freed. In the allocation process there needs to be a way to setup (3) commands and (4) polls.

The minimum parts to solve would be how to write engines in interpreted code. Something like this may work: https://gist.github.com/antonhornquist/7f4bb521e0eafa5ee4a180d07afe5db6 (a rwerite of Glut) or https://gist.github.com/antonhornquist/342acda2c9eae9ff5bf6dfb3d8bc0f39 (a larger example: Ack).

If `CroneEngine.load` was implemented with plumbing necessary for this to work it would have to be glued together with some startup logic like this: https://gist.github.com/antonhornquist/9b8be1f93afc66f81dc121841ceafb30

**Comment by antonhornquist** [comment]
To be honest, though, my first idea around this was to take the first steps towards supporting multiple engine backends and introduce the lua metadata file I've proposed elsewhere (from which metadata on commands, polls would be read rather than received via osc).

With this file and some plumbing in the core norns engine load logic, only one SC engine at a time would become loaded. The `CroneEngine.load` variant with `startup.scd` loading all engines seems a bit heavy weight.

**Comment by catfact** [comment]
proposed structure looks really good. 

requiring metadata file seems like a good near-future step,

incrementally, i think we should extend / alter the engine load mechanism to find `.scd` on launch.

one possibility to retain backwards compat:
-  add a path parameter to `/engine/load/name`, representing current "working directory" of the script. 
- crone can check if `Engine_{name}` is a class name, and/or search for `{path}/lib/{name}.scd` or something

or similarly, add a new OSC method and have `engine.lua` figure out which one to use.

**Comment by tehn** [comment]
(sorry for my lack of sc knowledge)

would this mean people could simply include a duplicate `.scd` file (version locked, basically) in their script lib folder?

this approach seems like it'd solve a lot of collision problems.

**Comment by catfact** [comment]
yes. but let's be clear if that's really the recommended best practice.
